### PR TITLE
running Dynamo from command line and tests

### PR DESCRIPTION
**Pull request by holyjewsus**
### Purpose

This PR modifies Program.CS and adds a series of command flags that let a user automate the running of dynamo graphs and the setting of the graph to preset states.

I've modified this PR according to comments below.
* commandLineArguments parsing is now handled by NDesk.Options which we are already include in our repo and which handles / and - command line flags
*  Startup logic, model creation, preloading methods, and locale setting is now handled in Dynamo.Applications namespace
* Dynamo Sandbox and DynamoCLI both reference DynamoApplications to handle these startup activities.
* I've added DynamoCLI and DynamoApplications to the Mono.2013 solution and tested DynamoCLI on mac (it requires being run as sudo as it might create user folders to hold settings)

please see this proposed documentation for more info on available flags:
https://github.com/holyjewsus/Dynamo/wiki/Proposed-Command-Line-Documentation

see a video here:
https://www.dropbox.com/s/3k54qg4udbdmaqz/command%20line%20restorestate.mov?dl=0

### Impact assessment

- [x] Changes the way 3rd party nodes work (Add ikeough to the reviewer list)
- [x] Affects which components of the system depends on which other components (Add pboyer to the reviewer list)

* Link to communications doc (same doc as above): https://github.com/holyjewsus/Dynamo/wiki/Proposed-Command-Line-Documentation

#### Testing strategy

the commandLineRunner class separates out of program.cs the logic for invoking run operations on a DynamoModel, via an array of arguments. There are a basic tests in this PR that check each command flag and run the DynamoTestModelBase model with the command line runner.

There are also tests that invoke the main method of the DynamoCLI with the command line runner class.

### Reviewers

@ikeough 
@pboyer 

### FYIs

@jnealb 
@Benglin

### CODE REVIEW

**Comment by sharadkjaiswal** [comment]
FYI, @Benglin will this change have any impact on Dynamo studio command line options?

**Comment by pboyer** [comment]
I'll investigate once we get the `Presets` code in.

However, I'd highly recommend using the very excellent [Command Line Parser Library](https://commandline.codeplex.com/) for this instead of rolling your own.  

Out of the box, it provides very nice command line documentation, is highly readable, and has many years of support.

File: src/DynamoCore/Core/CommandLineRunner.cs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+using Dynamo.Models;
+using System.Xml;
+using System.Threading;
+using System.IO;
+
+namespace Dynamo.Core
+{
+    public struct CommandLineArguments
+    {
+        internal static CommandLineArguments FromArguments(string[] args)
```
Comment (pboyer) [inline src/DynamoCore/Core/CommandLineRunner.cs]: Yes, definitely use the Command Line Parser Library for this.  It will make your life much nicer.  Remember to include the license along with it, though.

Comment (pboyer) [inline src/DynamoCore/Core/CommandLineRunner.cs]: AFAIK `DynamoCore` probably should not have any knowledge of `CommandLineArguments`.  Move this code (translated to the Command Line Parser equivalent) to `DynamoSandbox` in a file called `Options.cs`.

File: src/DynamoSandbox/Program.cs [hunk 2]
```diff
@@ -154,6 +104,27 @@ internal class Program
     {
         private static SettingsMigrationWindow migrationWindow;
 
+        private static DynamoModel makeModel()
```
Comment (pboyer) [inline src/DynamoSandbox/Program.cs]: Always use `PascalCase` when writing .NET code:

https://msdn.microsoft.com/en-us/library/x2dbyw72(v=vs.71).aspx

Also, don't create duplicate code.  It looks like your method could be used in `MakeStandaloneAndRun`.

**Comment by pboyer** [comment]
FWIW, do you ever plan on running this on a non-windows platform?  `DynamoSandbox` is windows specific.  You might consider creating a separate executable project, `DynamoCLI`, instead of reusing `DynamoSandbox`.

**Comment by holyjewsus** [comment]
@pboyer the Command Line Parser library does not seem to support /x command line flags, if it is difficult to add support for them, then there might be a lot of fallout from that change, (Dynamo studio's locale command line flag, and /c for running recorded files in Dynamo in many tests.

I'll look into this further and check in with @Benglin

**Comment by ikeough** [comment]
@holyjewsus Here's what I recommend:

- Create a new console application project which creates `DynamoCLI.exe` 
- Create a new class library project which creates `DynamoApplications.dll`.
- Create a new `Dynamo.Applications` namespace.
- Move the following logic into static methods in `Applications.StartupUtils`:
  - Model creation (you can leave view creation logic in DynamoSandbox)
  - Command line parsing
  - Locale setting
  - Preloading.
- Reference `DynamoApplications.dll` in both `DynamoSandbox` and `DynamoCLI`.
- Refactor `DynamoSandbox` to use utility methods from `Dynamo.Application.StartupUtils`.
- Write startup logic in `DynamoCLI` using `Dynamo.Application.StartupUtils`.

The goal is to have an assembly that has shared logic for startup operations which is used by both Dynamo applications. @pboyer will most likely have some comments on this as well.

File: src/DynamoApplications/StartupUtils.cs [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+
+namespace Dynamo.Applications.StartupUtils
+{
+    internal class PathResolver : IPathResolver
+    {
+        private readonly List<string> additionalResolutionPaths;
+        private readonly List<string> additionalNodeDirectories;
+        private readonly List<string> preloadedLibraryPaths;
+
+        public PathResolver(string preloaderLocation, IEnumerable<string> librariesToPreload)
```
Comment (holyjewsus) [inline src/DynamoApplications/StartupUtils.cs]: I've modified the signature for PathResolver to take a ienumerable<string> or libraries so we can pass the specific libraries to load instead of implementing a new path resolver subclass

File: src/DynamoApplications/StartupUtils.cs [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+            regKey = regKey.OpenSubKey(regKey64);
+
+            //Get "InstallLocation" value as string for all the subkey that starts with "Dynamo"
+            return regKey.GetSubKeyNames().Where(s => s.StartsWith("Dynamo")).Select(
+                (s) => regKey.OpenSubKey(s).GetValue("InstallLocation") as string);
+        }
+    }
+    //this class is left unimplemented,unclear how to
+    //lookup installation locations on nix/mac
+    internal class CLILookUp : DynamoLookUp
```
Comment (holyjewsus) [inline src/DynamoApplications/StartupUtils.cs]: currently unimplemented

File: src/DynamoApplications/StartupUtils.cs [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+                LibraryVersion.Version221
+            };
+
+            var preloader = new Preloader(rootFolder, versions);
+            preloader.Preload();
+            geometryFactoryPath = preloader.GeometryFactoryPath;
+            preloaderLocation = preloader.PreloaderLocation;
+        }
+
+        public static DynamoModel MakeModel(bool CLI, IEnumerable<string> librariesToPreload)
```
Comment (holyjewsus) [inline src/DynamoApplications/StartupUtils.cs]: DynamoModel Constructor now takes a bool for CLI to control the loading path to attempt to load CLI without all dependencies, like the update manager etc.

same list of libraries to preload is added here to pass to path resolver constructor

File: src/DynamoCLI/Program.cs [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+using Dynamo.Models;
+using Dynamo.Services;
+using System.Text.RegularExpressions;
+
+namespace DynamoCLI
+{
+    internal class Program
+    {
+
+        protected static CommandLineArguments commandstringToArgs(string commandstring)
```
Comment (holyjewsus) [inline src/DynamoCLI/Program.cs]: copy paste erorr- will remove these two methods (belong in CLItests)

**Comment by holyjewsus** [comment]
@ikeough @pboyer I've updated this PR and added some line notes. In addition to the reset of the review I want to make sure that the changes I've made to path resolver's constructor are appropriate.

File: src/DynamoApplications/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,12 @@
+﻿using System.Reflection;
```
Comment (ikeough) [inline src/DynamoApplications/Properties/AssemblyInfo.cs]: Ensure that your new projects are using `AssemblySharedInfo.cs`. This will ensure correct version numbering.

File: src/DynamoApplications/StartupUtils.cs [hunk 8]
```diff
@@ -0,0 +1,302 @@
...
+                "ProtoGeometry.dll",
+                "DSCoreNodes.dll",
+                "DSOffice.dll",
+                "DSIronPython.dll",
+                "FunctionObject.ds",
+                "Optimize.ds",
+                "DynamoConversions.dll",
+                "DynamoUnits.dll",
+                "Tessellation.dll",
+                "Analysis.dll"
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: You're missing `Display.dll`.

File: src/DynamoApplications/StartupUtils.cs [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+                      GeometryFactoryPath = geometryFactoryPath,
+                      StartInTestMode = true
+                  };
+
+            //if we are building a model for CLI mode, then we don't want to start an updateManager
+            //for now, building an updatemanager instance requires finding Dynamo install location
+            //which if we are running on mac os or *nix will use different logic then SandboxLookup 
+
+            UpdateManagerConfiguration umConfig = null;
+            if (!CLI)
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: Negative cases are difficult to read. You could remove this if statement by doing something in the config's construction like:
```c#
{
  ...
  config.UpdateManager = CLI ? null : InitializeUpdateManager()
  ...
}

private void InitializeUpdateManager()
{
  var cfg = UpdateManagerConfiguration.GetSettings(new SandboxLookUp());
  var um = new Dynamo.UpdateManager.UpdateManager(cfg);
}

File: src/DynamoApplications/StartupUtils.cs [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+            PreloadShapeManager(ref geometryFactoryPath, ref preloaderLocation);
+
+            var config = new DynamoModel.DefaultStartConfiguration()
+                  {
+                      PathResolver = new PathResolver(preloaderLocation, librariesToPreload),
+                      GeometryFactoryPath = geometryFactoryPath,
+                      StartInTestMode = true
+                  };
+
+            //if we are building a model for CLI mode, then we don't want to start an updateManager
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: Move these comments into doc comments on the `InitializeUpdateManager` method proposed below.

File: src/DynamoApplications/StartupUtils.cs [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+            }
+        }
+
+        public static List<string> CLILibraries
+        {
+            get
+            {
+                return new List<string> 
+                {
+                //"ProtoGeometry.dll",
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: The CLI doesn't load anything? If this is true then this list shouldn't exist.

File: src/DynamoApplications/StartupUtils.cs [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+
+            optionsSet.Parse(args);
+
+            //check for incompatabile parameters
+            if ((!string.IsNullOrEmpty(presetStateid) || (!string.IsNullOrEmpty(presetFile) || (!string.IsNullOrEmpty(verbose)))) && string.IsNullOrEmpty(openfilepath))
+            {
+                Console.WriteLine("you must supply a file to open if you want to load a preset or presetFile, or want to save an evaluation output ");
+                
+            }
+
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: Remove extra whitespace.

File: src/DynamoApplications/StartupUtils.cs [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+            //if we are building a model for CLI mode, then we don't want to start an updateManager
+            //for now, building an updatemanager instance requires finding Dynamo install location
+            //which if we are running on mac os or *nix will use different logic then SandboxLookup 
+
+            UpdateManagerConfiguration umConfig = null;
+            if (!CLI)
+            {
+                umConfig = UpdateManagerConfiguration.GetSettings(new SandboxLookUp());
+                Debug.Assert(umConfig.DynamoLookUp != null);
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: Is this your `Debug.Assert`? What action do we take if this condition is hit in the application? I'm not a fan of `Debug.Assert` because it just tells you something is wrong without actually controlling the behavior of the application.

File: src/DynamoApplications/StartupUtils.cs [hunk 14]
```diff
@@ -0,0 +1,302 @@
...
+                //"FunctionObject.ds",
+                //"VMDataBridge.dll",
+                };
+
+            }
+        }
+
+    }
+
+    public static class Locale
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: Let's put all of these utility methods in the `StartupUtils` class. They just get a bit wordy when you have to say `Dynamo.Applications.StartupUtils.Preloading.CLILibraries`. Let's just have:

`Dynamo.Applications.StartupUtils.SetLocale`
`Dynamo.Applications.StartupUtils.MakeDynamoModel`
`Dynamo.Applications.StartupUtils.PreloadShapeManager`
`Dynamo.Applications.StartupUtils.DefaultSandboxLibraries`

File: src/DynamoCLI/Program.cs [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+
+
+
+        [STAThread]
+        static internal void Main(string[] args)
+        {
+
+            try
+            {
+                var cmdLineArgs = CommandLineArguments.FromArguments(args);
```
Comment (ikeough) [inline src/DynamoCLI/Program.cs]: This doesn't really roll off the tongue. Rename `CommandLineArguments.FromArguments` to `CommandLineArguments.Parse`. This is also more similar to the `Parse` method available on c# types.

File: src/DynamoApplications/StartupUtils.cs (hunk 3 above)
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: Be careful! the `IPathResolver` interface was carefully designed to to allow your implementation to encapsulate all that was needed to define a `PathResolver`. With your strategy, we now have a couple of static lists with all the libaries to load in them called things like `SandboxLibraries` and `CLILibraries`. For someone trying to use your utility methods, understanding that you have to provide one of these lists (and knowing which one to use) is actually more difficult then just having to provide the correct `PathResolver` object.

I recommend keeping the preloaded libraries lists inside the `PathResolver` implementation and making two implementations:

`SandboxPathResolver`
`CLIPathResolver`

`PathResolver`s aren't that big. You can put them in another file called `PathResolvers.cs`

File: src/DynamoApplications/StartupUtils.cs [hunk 16]
```diff
@@ -0,0 +1,302 @@
...
+                Debug.Assert(umConfig.DynamoLookUp != null);
+                config.UpdateManager = new Dynamo.UpdateManager.UpdateManager(umConfig);
+                config.StartInTestMode = false;
+            }
+
+            var model = DynamoModel.Start(config);
+            return model;
+        }
+
+        public static List<string> SandBoxLibraries
```
Comment (ikeough) [inline src/DynamoApplications/StartupUtils.cs]: See notes above on creating a `SandboxPathResolver`.

**Comment by ikeough** [comment]
@holyjewsus Lots of comments. Sorry for the back and forth. But you're changing the way that the application is created :), so it's worth being overly thorough.

**Comment by ikeough** [comment]
For documentation, do a `-h` and paste the command line arguments into the wiki. It's amazing how many people don't use `-h` from the command line. Having a list of the available flags on the wiki is helpful.

File: test/Libraries/CommandLineTests/CommandLineTests.cs [hunk 17]
```diff
@@ -0,0 +1,171 @@
...
+        }
+
+        [Test]
+        public void CanOpenAndRunSpecifcStateAndOutputToFileFromDynamoCLIexe()
+        {
+            string openPath = Path.Combine(TestDirectory, @"core\commandline\addmultwithstatesindyn.dyn");
+            var newPath = GetNewFileNameOnTempPath("xml");
+            string commandstring = "/o" + " " + openPath + " " + "/s" + " " + "state2" + " " + "/v" + " " + newPath;
+
+            DynamoCLI.Program.Main(commandstringToStringArray(commandstring));
```
Comment (ikeough) [inline test/Libraries/CommandLineTests/CommandLineTests.cs]: https://msdn.microsoft.com/en-us/library/x2dbyw72(v=vs.71).aspx ;)

File: test/Libraries/CommandLineTests/CommandLineTests.cs [hunk 18]
```diff
@@ -0,0 +1,171 @@
...
+
+namespace Dynamo.Tests
+{
+   
+    internal class CommandLineTests : DynamoModelTestBase
+    {
+
+        protected override void GetLibrariesToPreload(List<string> libraries)
+        {
+            libraries.Add("VMDataBridge.dll");  // Required for Watch node.
```
Comment (ikeough) [inline test/Libraries/CommandLineTests/CommandLineTests.cs]: Do you need to load all of these libraries for your tests? Load only the bare minimum required to keep testing time to a minimum.

**Comment by holyjewsus** [comment]
@ikeough @pboyer , I've made the changes requested to the code.

The debug.assert was not mine, I've moved it into the private initializeUpdateManager method

I've reverted PathResolver to the original implementation

The list of libraries needed for the CLI might be dependent on what a user needs to run, for now I've limited it to core libraries, geometry, python, function application, watch nodes, and Units/conversion

**Comment by holyjewsus** [comment]
@pboyer the appVeyor failure confuses me, both 2012 and 2013 .slns build fine on my machine, the failure is OutputPath not set on DynamoPackages, but it appears to be in the csproj.

any hints?

**Comment by holyjewsus** [comment]
I am going to close this PR- and open a new PR, I think the appveyor failure is due to a force push I had to do to this PR branch.
